sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala

```diff
@@ -271,6 +271,8 @@ class ParquetToSparkSchemaConverter(
             } else {
               TimestampNTZType
             }
+          case timestamp: TimestampLogicalTypeAnnotation if timestamp.getUnit == TimeUnit.NANOS =>
```

Comment:
Ideally, this case would be merged with above case, but that would require `TimestampType` and `TimestampNTZType` to support nanos, which is a bigger change.

This `case` deserves a comment that nanos are not supported as `TimestampType` but as `LongType`, without any timezone awareness.

Supporting nanos as `TimestampType` in the future looks like a breaking change then (Spark 4.x?). Or another `TimestampNSType` like `TimstampNTZType` could be introduced.

Reply:
Comment has been added.

I agree, full support is a breaking change and my aim for now is to address the regression and not introduce new functionality/support as that would require further validation and testing of other components e.g. time based functions  etc. However, definitely something that should be considered in future development

Reply:
Hmm, how is this sufficient? we also should handle the long (nanos) accordingly in the Parquet reader, is that right? e.g.: `ParquetVectorUpdaterFactory`.

Reply:
@sunchao so for type such as

```
message root {
  required int64 _1 (TIMESTAMP(NANOS,true));
}
```

the `descriptor.getPrimitiveType().getPrimitiveTypeName()` is INT64; as the `sparkType` is `LongType` it returns `new LongUpdate();` 

https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java#L105

Therefore no change is required from this perspective as it continues to be handled in the same way prior to https://github.com/apache/spark/pull/31776

Reply:
I see. It's a bit weird though that we read nanos as long, but not timestamp. I'm not sure whether this should be considered as a regression, or that the previous behavior before 3.2 is merely unintended.

Reply:
Yes, the (earlier) existing behaviour was useful and should be restored until properly supported as typed nano timestamp. Unless a workaround can be found that restores the earlier behaviour.

Does providing a schema (`spark.read.schema(...)`) with long type override the parquet timestamp type from the parquet file? Would that be a workaround?

Reply:
@EnricoMi so it is possible to use `spark.read.schema(..)` as a workaround, however, you end up loosing functionality like `mergeSchema` which will automatically handle schema evolution etc and potentially will need to know the entire schema up front if all columns are required. Also for other consumers/users, especially in the exploratory analysis space, it will require them to have a better understanding of the underlying data structure before they can use it and this gets more difficult when the file is extremely wide.

I can imagine developers creating otherways to avoid the nuisance; which seems a bit crazy considering that functionality already exists.

Reply:
So this is not a valid workaround. @sunchao how is your feeling about restoring earlier useful behaviour?

Reply:
Make sense. I'm OK with it.